### ISSUE TITLE
Implement ability to enforce unofficial Bash script mode

### ISSUE THREAD

**Issue by btamayo** [bats-core/bats-core#36]
> Oh, one more thing @jasonkarns: Adding a feature to enforce this mode is an interesting idea. Wanna file it for v1.0.0 (or v1.1.0)?

> https://github.com/bats-core/bats-core/pull/23#issuecomment-336245484

**Comment by trnubo** [comment]
I think my use case is related to this. I'm using bats to test a set of bash functions instead of a simple bash script but I'd like to make sure the test all pass with bash strict mode. At the moment this doesn't seem possible unless I set strict mode in every function which is not what I want to do and setting strict mode at the top of my `functions.sh` doesn't seem to have an affect.

strict.bats
```
#!/usr/bin/env bats

source functions.sh

@test "Run false true" {
  run func_false_true
  [ "$status" -ne 0 ]
}

@test "Run strict false true" {
  run func_strict_false_true
  [ "$status" -ne 0 ]
}
```

functions.sh
```
# Nether of these should exit zero because the false under strict mode should force a return
set -euo pipefail
IFS=$'\n\t'

func_false_true() {
	false
	true
}

func_strict_false_true() {
	set -euo pipefail
	IFS=$'\n\t'
	false
	true
}
```

Results
```
 ✗ Run false true
   (in test file strict.bats, line 7)
     `[ "$status" -ne 0 ]' failed
 ✓ Run strict false true

2 tests, 1 failure
```

**Comment by tterranigma** [comment]
@trnubo The reason this does not work as expected is that Bats internally disables execfail in the [`run` function](https://github.com/bats-core/bats-core/blob/master/libexec/bats-exec-test#L55). 

Take a look at #69 for suggestions from @mbland on how to have this behavior in your tests.

**Comment by martin-schulze-vireso** [comment]
I read through lots of linked issues to gauge if this was fixed by #26 or if there is more to it.

My understanding is that this issue is about adding the strict mode (`set -euo pipefail`) into bats internal files, and not into a test as has been done in #26. Can anyone of the previously involved people put more detail into this, so we can decide if it is worth the effort or should be closed? @jasonkarns @sublimino

**Comment by sublimino** [comment]
You're correct @martin-schulze-vireso, #26 is about the bats test subshells, this issue is about the bats runner. From https://github.com/bats-core/bats-core/pull/23#issuecomment-336245484:

> @mbland I do think the warning might be better off in a separate PR. reason being... I'm beginning to see a theme around bats having its own "strict mode" so to speak. consider:
> 
> warning about empty test names

IMO this should be a fatal error

> warning about duplicate test names

This is now a fatal error to avoid potential nondeterministic invocation of one of the N duplicate test names

> bats itself being in or out-of UnofficialBashStrictMode

This is to enable strict mode for the bats runner (`set -Eeuxo pipefail` — note the `-E` to propagate to subshells). This should be always-on IMHO, although might break plugins?

> bats "fixing" known bash gotchas (like [[)

This would be a useful addition for the contents of bats tests themselves.

> Perhaps we should have a more comprehensive flag that enables bats "strict" mode (or maybe even better, defaults to strict mode and offer a --loose alternative). In this theoretical strict mode, empty and duplicate test names automatically fail; bats itself is placed in strict mode before the execution of every test (to ensure the sourced test doesn't disable it); and we automatically fix any bash gotchas that we know about?

IMHO defaulting bats itself to strict, and adding a flag for `--loose` / `--no-strict`  is preferable. That flag may be for the gotchas, and defaulting to strict mode for each test's subshell (both breaking changes).

**Comment by rockandska** [comment]
I upvote the fact that strict mode should be enforce everywhere and disabled only when needed.

Right now, i'm using this to achieve this purpose and seems ok for now :

```
setup() {
	set -Eeuo pipefail
}

run() {
	local origFlags="$-"
	local origIFS="$IFS"
	local previous_trap
	local output_file
	local rc
	printf -v "previous_trap" '%s\n' "$(trap)"
	trap 'rc=$?; trap - ERR;return 0' ERR
	output_file=$(mktemp "${BATS_TMPNAME}.fifo.XXXXXX")
	set -Eeuo pipefail
	"$@" > "${output_file}" 2>&1
	set +Eeuo pipefail
	eval "${previous_trap:-}"
	status="${rc:-0}"
	output=$(< "${output_file}")
	IFS=$'\n' read -rd '' -a lines <<<"${output}" || true
	IFS="$origIFS"
	set "-$origFlags"
}
```

- if a unbound variable is detected it throw an error (not possible to catch unbound variable)
- if an error occur with `run` command, the ERR trap catch the exit status but do not exit to let the user test the status / output / lines as usual

Tested with bash 3.2.57, 4.1.17 , 4.4.23
